packages/cosmic-swingset/lib/ag-solo/vats/lib-dehydrate.js

```diff
@@ -0,0 +1,82 @@
+import harden from '@agoric/harden';
+import { makeMarshal } from '@agoric/marshal';
+import makeStore from '@agoric/store';
+import { assert, details } from '@agoric/assert';
+
+// Marshalling for the UI should only use the user's petnames. We will
+// call marshalling for the UI "dehydration" and "hydration" to distinguish it from
+// other marshalling.
+export const makeDehydrator = (initialUnnamedCount = 0) => {
+  let unnamedCount = initialUnnamedCount;
+
+  const petnameKindToMappings = makeStore();
+
+  const searchOrder = [];
+
+  const makeMappings = kind => {
+    assert.typeof(kind, 'string', details`kind ${kind} must be a string`);
+    const valToPetname = makeStore();
+    const petnameToVal = makeStore();
+    const addPetname = (petname, val) => {
+      petnameToVal.init(petname, val);
+      valToPetname.init(val, petname);
```

Comment:
Nice use of `init` to cleanly reject collisions without extra mechanism.

Reply:
This collision rejection is the bug I mentioned in the meeting this morning. Depending on previous bindings, it is possible for the first `init` to succeed, because the `petname` is new, but the second `init` to throw because the `val` had already been associated with a different `petname`. This leave the mapping (aka "mappings") in an inconsistent state where two petnames map to the same `val`.

Reply:
Good catch, thanks!